[LT] Move device lock in LazyGraphExecutor to a later place

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76974

Summary: cherry-picking https://github.com/pytorch/xla/pull/3457

desertfire committed:
[LT] Move device lock in LazyGraphExecutor to a later place

Summary: cherry-picking https://github.com/pytorch/xla/pull/3457

[ghstack-poisoned]

wconstab reviewed (approved):
lgtm, thanks bin!

JackCaoG commented on torch/csrc/lazy/core/lazy_graph_executor.cpp:
```diff
@@ -1096,12 +1098,30 @@ std::vector<BackendDataPtr> LazyGraphExecutor::GatherTensorsData(
 }
 
+void LazyGraphExecutor::TensorCollectionBarrier(SyncTensorCollection* coll) {
+  static const std::string invalid_device(
+      "Unknown0"); /* Temp solution to idetify unassigned devices */
```
just a note for myself to remove this after https://github.com/pytorch/pytorch/pull/76517/files merged and each backend implement the default device correctly.

JackCaoG reviewed (approved):
Thanks Bin!

desertfire committed:
Update on "[LT] Move device lock in LazyGraphExecutor to a later place"


Summary: cherry-picking https://github.com/pytorch/xla/pull/3457

[ghstack-poisoned]

desertfire committed:
Update on "[LT] Move device lock in LazyGraphExecutor to a later place"


Summary: cherry-picking https://github.com/pytorch/xla/pull/3457

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge this
